File: src/Actor.GameHub.Identity.Actors/IdentityActor.cs
using System;
using System.Collections.Generic;
using Actor.GameHub.Identity.Abstractions;
using Akka.Actor;
using Akka.Event;

namespace Actor.GameHub.Identity.Actors
{
  public class IdentityActor : ReceiveActor
  {
    private readonly ILoggingAdapter _logger = Context.GetLogger();

    private readonly Dictionary<Guid, (LoginUserMsg LoginMsg, IActorRef LoginOrigin)> _loginOriginByAuthId = new();
    private readonly Dictionary<IActorRef, Guid> _authIdByAuthenticatorRef = new();

    public IdentityActor()
    {
      Receive<LoginUserMsg>(msg => !msg.IsValid(), LoginInvalid);
      Receive<LoginUserMsg>(LoginUser);
      Receive<UserAuthErrorMsg>(AuthError);
      Receive<UserAuthSuccessMsg>(AuthSuccess);
      Receive<Terminated>(OnTerminated);

      _logger.Info("==> Identity started");
    }

    private void LoginInvalid(LoginUserMsg loginMsg)
    {
      Sender.Tell(new UserLoginErrorMsg
      {
        UserLoginId = loginMsg.UserLoginId,
        ErrorMessage = "invalid login",
      });
    }

    private void LoginUser(LoginUserMsg loginMsg)
    {
      var authUserMsg = new AuthUserMsg
      {
        AuthId = Guid.NewGuid(),
        LoginUserMsg = loginMsg,
      };

      if (_loginOriginByAuthId.TryAdd(authUserMsg.AuthId, (loginMsg, Sender)))
      {
        var authenticator = Context.ActorOf(UserAuthenticatorActor.Props(), IdentityMetadata.UserAuthenticatorName(authUserMsg.AuthId));
        _authIdByAuthenticatorRef.Add(authenticator, authUserMsg.AuthId);
        Context.Watch(authenticator);
        authenticator.Tell(authUserMsg);
      }
      else
      {
        var loginErrorMsg = new UserLoginErrorMsg
        {
          UserLoginId = loginMsg.UserLoginId,
          ErrorMessage = "user login authId error, try again...",
        };
        Sender.Tell(loginErrorMsg);
      }
    }

    private void AuthError(UserAuthErrorMsg authErrorMsg)
    {
      var authenticatorRef = Sender;

      if (_loginOriginByAuthId.TryGetValue(authErrorMsg.AuthId, out var data)
        && _authIdByAuthenticatorRef.ContainsKey(authenticatorRef))
      {
        var loginErrorMsg = new UserLoginErrorMsg
        {
          UserLoginId = data.LoginMsg.UserLoginId,
          ErrorMessage = authErrorMsg.ErrorMessage,
        };
        data.LoginOrigin.Tell(loginErrorMsg);

        _loginOriginByAuthId.Remove(authErrorMsg.AuthId);
        _authIdByAuthenticatorRef.Remove(authenticatorRef);
        Context.Stop(authenticatorRef);
      }
    }

    private void AuthSuccess(UserAuthSuccessMsg authSuccessMsg)
    {
      var authenticatorRef = Sender;

      if (_loginOriginByAuthId.TryGetValue(authSuccessMsg.AuthId, out var data)
        && _authIdByAuthenticatorRef.ContainsKey(authenticatorRef))
      {
        var loginSuccessMsg = new UserLoginSuccessMsg
        {
          UserLoginId = data.LoginMsg.UserLoginId,
          User = authSuccessMsg.User,
        };
        data.LoginOrigin.Tell(loginSuccessMsg);

        _loginOriginByAuthId.Remove(authSuccessMsg.AuthId);
        _authIdByAuthenticatorRef.Remove(authenticatorRef);
        Context.Stop(authenticatorRef);
      }
    }

    private void OnTerminated(Terminated terminatedMsg)
    {
      var authRef = terminatedMsg.ActorRef;

      if (_authIdByAuthenticatorRef.TryGetValue(authRef, out var authId)
        && _loginOriginByAuthId.TryGetValue(authId, out var data))
      {
        _logger.Warning($"{nameof(OnTerminated)}: unexpected stop of authenticator {authId}, {authRef.Path}");
        _authIdByAuthenticatorRef.Remove(authRef);
        _loginOriginByAuthId.Remove(authId);

        var loginErrorMsg = new UserLoginErrorMsg
        {
          ErrorMessage = "user login error, unexpected stop of authenticator",
        };
        data.LoginOrigin.Tell(loginErrorMsg);
      }
    }

    public static Props Props()
      => Akka.Actor.Props.Create<IdentityActor>();
  }
}

File: fun/Actor.GameHub.Identity.FSharp/IdentityActor.fs
module Actor.GameHub.Identity.IdentityActor

open System
open Akka.FSharp
open Akka.Actor

open Actor.GameHub.Identity.Abstractions

type AuthData =
    { LoginId: Guid
      LoginOrigin: IActorRef }

type IdentityState =
    { AuthDataByAuthId: Map<Guid, AuthData> }

let initialIdentityState = { AuthDataByAuthId = Map.empty }

let addAuth authId loginId loginOrigin state =
    { state with
          AuthDataByAuthId =
              Map.add
                  authId
                  { LoginId = loginId
                    LoginOrigin = loginOrigin }
                  state.AuthDataByAuthId }

let removeAuth authId state =
    { state with
          AuthDataByAuthId = Map.remove authId state.AuthDataByAuthId }

let handleLogin spawnAuthenticator loginId username (mailbox: Actor<_>) state =
    let loginOrigin = mailbox.Sender()
    let authId = Guid.NewGuid()

    match Map.containsKey authId state.AuthDataByAuthId with
    | true ->
        loginOrigin
        <! UserLoginErrorMsg(loginId, "user loginId error, try again...")

        state
    | false ->
        let authUserMsg =
            AuthUserMsg(authId, { Username = username })

        let authRef =
            spawnAuthenticator mailbox.Context authId

        mailbox.Context.WatchWith(authRef, AuthTerminated authId)
        <! authUserMsg

        addAuth authId loginId loginOrigin state

let handleAuthError authId errorMessage (mailbox: Actor<_>) state =
    let authRef = mailbox.Sender()

    match Map.tryFind authId state.AuthDataByAuthId with
    | None -> state
    | Some loginData ->
        loginData.LoginOrigin
        <! UserAuthErrorMsg(loginData.LoginId, errorMessage)

        mailbox.Unwatch(authRef) |> ignore
        mailbox.Context.Stop(authRef)
        removeAuth authId state

let handleAuthSuccess authId user (mailbox: Actor<_>) state =
    let authRef = mailbox.Sender()

    match Map.tryFind authId state.AuthDataByAuthId with
    | None -> state
    | Some loginData ->
        loginData.LoginOrigin
        <! UserLoginSuccessMsg(loginData.LoginId, user)

        mailbox.Unwatch(authRef) |> ignore
        mailbox.Context.Stop(authRef)
        removeAuth authId state

let handleAuthTerminated authId state =
    match Map.tryFind authId state.AuthDataByAuthId with
    | None -> state
    | Some authData ->
        authData.LoginOrigin
        <! UserLoginErrorMsg(authData.LoginId, $"unexpected stop of authenticator {authId}")

        removeAuth authId state

let identity spawnAuthenticator (mailbox: Actor<_>) =
    let rec loop state =
        actor {
            let! msg = mailbox.Receive()

            let nextState =
                match msg with
                | LoginUserMsg (loginId, username) -> handleLogin spawnAuthenticator loginId username mailbox
                | UserAuthErrorMsg (authId, errorMessage) -> handleAuthError authId errorMessage mailbox
                | UserAuthSuccessMsg (authId, user) -> handleAuthSuccess authId user mailbox
                | AuthTerminated (authId) -> handleAuthTerminated authId

            return! loop (nextState state)
        }

    loop initialIdentityState

let spawnIdentity spawnAuthenticator system =
    spawn system IdentityMetadata.IdentityName (identity spawnAuthenticator)
